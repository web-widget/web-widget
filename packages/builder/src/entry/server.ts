// @ts-expect-error
import userManifest from "virtual:@web-widget/server:user-manifest";

import type { Manifest, Router } from "@web-widget/web-server";
import webServer from "@web-widget/web-server";
import { createWebRequest, sendWebResponse } from "@web-widget/express";
import { IncomingMessage, ServerResponse } from "node:http";

async function resolveManifest(
  object: { module: string }[] | { module: string }
) {
  const result: Record<string, any> = {};

  for (const [key, value] of Object.entries(object)) {
    if (Array.isArray(value)) {
      result[key] = await Promise.all(
        value.map(({ module: file, ...value }) =>
          import(/* @vite-ignore */ file).then((module) => ({
            ...value,
            module,
          }))
        )
      );
    } else {
      result[key] = await import(/* @vite-ignore */ value.module).then(
        (module) => ({
          ...value,
          module,
        })
      );
    }
  }

  return result as Manifest;
}

let router: Router;
const manifest = resolveManifest(userManifest);

export const middleware = async (
  req: IncomingMessage,
  res: ServerResponse,
  callback: (...args: unknown[]) => void
) => {
  if (!router) {
    router = webServer(await manifest);
  }

  const webRequest = createWebRequest(req, res);
  const webResponse = await router.handler(webRequest);

  // TODO
  // @ts-ignore
  await sendWebResponse(res, webResponse);
};
