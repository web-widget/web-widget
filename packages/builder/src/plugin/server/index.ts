import type { Manifest } from "@web-widget/web-server";
import type { Plugin } from "vite";
import { createViteLoader, ModuleLoader } from "../../core/loader/index";
import { handleRequest } from "./request";

export default function createVitePluginServer(): Plugin {
  async function getManifest(loader: ModuleLoader) {
    return loader.import("/web-server.gen.ts") as Promise<Manifest>;
  }
  return {
    name: "@web-widget/builder:server",
    async configureServer(viteServer) {
      const loader = createViteLoader(viteServer);
      let manifest: Manifest | void;
      let pendingReload: DeferredPromise<Manifest> | void;

      /** rebuild the route cache + manifest, as needed. */
      async function rebuildManifest(
        needsManifestRebuild: boolean,
        _file: string
      ) {
        if (needsManifestRebuild) {
          pendingReload = new DeferredPromise<Manifest>()
          manifest = await getManifest(loader);
          pendingReload.complete(manifest)
          pendingReload = undefined;
        }
      }
      // Rebuild route manifest on file change, if needed.
      viteServer.watcher.on("add", rebuildManifest.bind(null, true));
      viteServer.watcher.on("unlink", rebuildManifest.bind(null, true));
      viteServer.watcher.on("change", rebuildManifest.bind(null, true));

      return async () => {
        // Note that this function has a name so other middleware can find it.
        viteServer.middlewares.use(async function widgetServerDevHandler(
          req,
          res
        ) {
          if (pendingReload) await pendingReload.p;
          manifest = manifest || (await getManifest(loader));
          return handleRequest(manifest, viteServer, loader, req, res);
        });
      };
    },
  };
}

export type ValueCallback<T = unknown> = (value: T | Promise<T>) => void;

export class DeferredPromise<T> {
  private completeCallback!: ValueCallback<T>;

  public readonly p: Promise<T>;

  constructor() {
    this.p = new Promise<T>((c) => {
      this.completeCallback = c;
    });
  }

  public complete(value: T) {
    return new Promise<void>(resolve => {
      this.completeCallback(value);
      resolve();
    });
  }
}
