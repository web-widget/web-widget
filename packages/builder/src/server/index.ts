import type { ModuleLoader } from "./loader/index";
import { createViteLoader } from "./loader/index";
import type {
  Plugin,
  ServerOptions,
  UserConfig as ViteUserConfig,
  ViteDevServer,
} from "vite";
import { createServer as createViteServer, mergeConfig } from "vite";
import { join, relative } from "node:path";
import { openConfig, resolveConfigPath } from "../config";

import type { BuilderConfig } from "../types";
import type { ManifestJSON } from "@web-widget/web-router";
import { fileURLToPath } from "node:url";
import fs from "node:fs";
import pc from "picocolors";
import stripAnsi from "strip-ansi";

import type { Middleware } from "@web-widget/node";
import NodeAdapter from "@web-widget/node";

import WebRouter from "@web-widget/web-router";
import { getMeta } from "./render";
import type { RouteModule } from "@web-widget/schema";

async function loadManifest(
  config: BuilderConfig,
  loader: ModuleLoader
): Promise<ManifestJSON> {
  const json = await loader.import(
    join(
      config.base,
      relative(fileURLToPath(config.root), fileURLToPath(config.input))
    )
  );
  return json.default as ManifestJSON;
}

function getWatchFiles(config: BuilderConfig, manifest: ManifestJSON) {
  const modules: string[] = [];

  Array.from(Object.entries(manifest)).forEach(([key, value]) => {
    if (Array.isArray(value)) {
      value.forEach(({ module }) => {
        modules.push(join(fileURLToPath(config.root), module));
      });
    } else if (value.module) {
      modules.push(join(fileURLToPath(config.root), value.module));
    }
  });

  return modules;
}

function createVitePluginServer(config: BuilderConfig): Plugin {
  return {
    name: "@web-widget/builder:server",
    async configureServer(viteServer) {
      const configPath = (await resolveConfigPath({
        cwd: fileURLToPath(config.root),
        fs,
      })) as string;
      const loader = createViteLoader(viteServer);
      const manifest = await loadManifest(config, loader);
      const watchFiles: string[] = getWatchFiles(config, manifest);

      async function restart(file: string) {
        if (
          file === configPath ||
          file === fileURLToPath(config.input.href) ||
          watchFiles.includes(file)
        ) {
          viteServer.config.logger.info(
            pc.green(
              `${relative(
                fileURLToPath(config.root),
                file
              )} changed, restarting server...`
            ),
            { clear: true, timestamp: true }
          );
          await viteServer.close();
          // @ts-ignore
          global.__vite_start_time = Date.now();
          const { viteServer: newServer } = await createServer(
            viteServer.config.root,
            viteServer.config.server
          );
          await newServer.listen();
        }
      }
      // Rebuild route manifest on file change, if needed.
      viteServer.watcher.on("add", restart);
      viteServer.watcher.on("change", restart);

      const webRouterDevMiddleware = toWebRouterDevMiddleware(
        config,
        manifest,
        viteServer,
        loader
      );

      return () => {
        viteServer.middlewares.use(webRouterDevMiddleware);
      };
    },
  };
}

export async function createServer(
  root: string = process.cwd(),
  serverOptions: ServerOptions = {}
) {
  const { builderConfig } = await openConfig({
    cwd: process.cwd(),
    cmd: "server",
    // mode: 'dev'
  });
  const viteConfig = mergeConfig(builderConfig.vite, {
    root,
    appType: "custom",
    logLevel: "info",
    plugins: [createVitePluginServer(builderConfig)],
    server: serverOptions,
  } as ViteUserConfig);

  return {
    //config,
    viteConfig,
    viteServer: await createViteServer(viteConfig),
  };
}

function toWebRouterDevMiddleware(
  config: BuilderConfig,
  manifest: ManifestJSON,
  viteServer: ViteDevServer,
  loader: ModuleLoader
): Middleware {
  const webRouter = new WebRouter(manifest, {
    dev: true,
    baseAsset:
      (viteServer.resolvedUrls?.network ||
        viteServer.resolvedUrls?.local)?.[0] ?? config.base,
    baseModule: config.root,
    experimental: {
      loader: async (id, importer) => {
        const source = new URL(id, importer);
        const module = {
          meta: {},
          ...((await loader.import(fileURLToPath(source))) as RouteModule),
        };

        const meta = await getMeta(source, loader, config.root, "development");
        Object.entries(meta).forEach(([key, value]) => {
          // @ts-ignore
          if (module.meta[key]) {
            // @ts-ignore
            module.meta[key].push(...value);
          } else {
            // @ts-ignore
            module.meta[key] = value;
          }
        });

        return module;
      },
    },
  });

  const nodeAdapter = new NodeAdapter({
    async handler(request, fetchEvent) {
      try {
        const webResponse = await webRouter.handler(request, fetchEvent);

        if (
          !webResponse.headers.get("content-type")?.startsWith("text/html;")
        ) {
          return webResponse;
        }

        const html = await webResponse.text();
        const viteHtml = await viteServer.transformIndexHtml(
          new URL(request.url).pathname,
          html
        );

        return new Response(viteHtml, {
          status: webResponse.status,
          statusText: webResponse.statusText,
          headers: webResponse.headers,
        });
      } catch (error) {
        viteServer.ssrFixStacktrace(error);
        console.error(error.stack);

        return new Response(errorTemplate(stripAnsi(error.stack)), {
          status: 500,
          statusText: "Internal Server Error",
          headers: {
            "content-type": "text/html; charset=utf-8",
          },
        });
      }
    },
  });

  return nodeAdapter.middleware;
}

function errorTemplate(message: string) {
  return `<!doctype html>
  <html lang="en">
    <head>
      <title>Error</title>
      <meta name="viewport" content="width=device-width, initial-scale=1.0">
    </head>
    <body>
      <div style="display:flex;justify-content:center;align-items:center">
        <div style="border:#f3f4f6 2px solid;border-top:red 4px solid;background:#f9fafb;margin:16px;min-width:550px">
          <p style="margin:0;font-size:12pt;padding:16px;font-family:sans-serif"> An error occurred during route handling or page rendering. </p>
          <pre style="margin:0;font-size:12pt;overflow-y:auto;padding:16px;padding-top:0;font-family:monospace">${message}</pre>
        </div>
      </div>
    <body>
  </html>`;
}
