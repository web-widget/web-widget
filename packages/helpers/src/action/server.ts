export * from '@web-widget/action/server';
