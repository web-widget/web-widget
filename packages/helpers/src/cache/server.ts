export * from '@web-widget/lifecycle-cache/server';
