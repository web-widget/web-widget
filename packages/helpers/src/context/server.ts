export * from '@web-widget/context/server';
export * from './state';
