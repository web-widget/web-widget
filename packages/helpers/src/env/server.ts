export const IS_CLIENT = false;
export const IS_SERVER = true;
