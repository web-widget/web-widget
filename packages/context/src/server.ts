import type { MiddlewareContext, ScriptDescriptor } from '@web-widget/schema';
import { tryGetAsyncLocalStorage } from './context';
import { SCRIPT_ID, EXPOSED_TO_CLIENT } from './constants';
export const exposedToClient = EXPOSED_TO_CLIENT;

export function contextToScriptDescriptor(
  context: MiddlewareContext
): ScriptDescriptor {
  return {
    id: SCRIPT_ID,
    type: 'application/json',
    content: JSON.stringify({
      params: context.params,
    }),
  };
}

export function callContext<T extends (...args: any[]) => any>(
  data: MiddlewareContext,
  setup: T,
  args?: Parameters<T>
) {
  const ctx = tryGetAsyncLocalStorage<MiddlewareContext>();
  const fn: () => ReturnType<T> = () =>
    args ? setup(...(args as Parameters<T>)) : setup();
  return ctx.call(data, fn);
}

export function context() {
  const ctx = tryGetAsyncLocalStorage();
  return ctx.use() as MiddlewareContext;
}

/** @deprecated Use `context` instead. */
export const useContext = context;
