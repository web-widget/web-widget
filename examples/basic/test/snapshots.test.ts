import routes from '../routemap.server.json';
import { createTestServer, type Server } from './server';

let server: Server;
const STATIC_ROUTES = routes.routes.filter(
  ({ pathname }) => !/[:(){}*+?]/.test(pathname)
);

beforeAll(async () => {
  server = await createTestServer();
});

afterAll(async () => {
  await server.close();
});

describe('Should match snapshot', () => {
  test.each(STATIC_ROUTES.map((route) => [route.pathname]))(
    'Request "%s" should match snapshot',
    async (pathname) => {
      const result = await server.fetch(`${pathname}`);
      expect(result.status).toMatchSnapshot(`${pathname}@status`);
      expect(result.statusText).toMatchSnapshot(`${pathname}@statusText`);
      expect(result.headers).toMatchSnapshot(`${pathname}@headers`);
      expect(await result.text()).toMatchSnapshot(`${pathname}@body`);
    }
  );
});
