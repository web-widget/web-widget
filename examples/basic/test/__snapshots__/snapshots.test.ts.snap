// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should match snapshot Request "/" should match snapshot: /@body 1`] = `
"<!doctype html>
    <html lang="en">
      <head>
        <script type="module">
import RefreshRuntime from "/@react-refresh"
RefreshRuntime.injectIntoGlobalHook(window)
window.$RefreshReg$ = () => {}
window.$RefreshSig$ = () => (type) => type
window.__vite_plugin_react_preamble_installed__ = true
</script>

        <script type="module" src="/@vite/client"></script>

        <script id="shim:es-module">
  if (!HTMLScriptElement.supports || !HTMLScriptElement.supports('importmap')) {
    self.importShim = Object.assign(
      function importShimProxy() {
        const esModuleShimUrl =
          'https://ga.jspm.io/npm:es-module-shims@1.7.3/dist/es-module-shims.js';
        const promise = new Promise((resolve, reject) => {
          document.head.appendChild(
            Object.assign(document.createElement('script'), {
              src: esModuleShimUrl,
              crossorigin: 'anonymous',
              async: true,
              onload() {
                if (!importShim.$proxy) {
                  resolve(importShim);
                } else {
                  reject(
                    new Error(
                      'No globalThis.importShim found:' + esModuleShimUrl
                    )
                  );
                }
              },
              onerror(error) {
                reject(error);
              },
            })
          );
        });
        return promise.then((importShim) => importShim(...arguments));
      },
      {
        $proxy: true,
      }
    );
  }
</script> <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1.0" /><title >Home - Web Widget</title><script id="state:web-router" type="application/json">{"pathname":"/","params":{},"body":{}}</script>
      <link rel="stylesheet" href="/routes/(components)/Counter.module.css" /><link rel="stylesheet" href="/routes/(components)/BaseLayout.module.css" /><style data-vite-dev-id="#TEST_WORKSPACE_DIR#/examples/basic/routes/(components)/Counter@widget.vue?vue&amp;type=style&amp;index=0&amp;scoped=a67ee2b2&amp;lang.css">
.counter[data-v-a67ee2b2] {
  display: inline-block;
  line-height: 1em;
  padding: 15px;
  border-radius: 30px;
  font-size: 16px;
  border: 2px solid #42b883;
}
.counter button[data-v-a67ee2b2] {
  width: 2em;
  height: 2em;
  line-height: 2em;
  text-align: center;
  border-radius: 100%;
  font-size: 1em;
  border: 0 none;
  outline: #fff solid 5px;
  background: #42b883;
  color: #fff;
  font-weight: bold;
  cursor: pointer;
}
.counter .count[data-v-a67ee2b2] {
  display: inline-block;
  min-width: 1.2em;
  text-align: center;
  padding-left: 10px;
  padding-right: 10px;
  font-weight: bold;
  font-size: 1em;
  color: #42b883;
  border: 1px solid transparent;
  background: transparent;
}
</style><script type="module" src="/routes/(components)/Counter@widget.vue?vue&amp;type=style&amp;index=0&amp;scoped=a67ee2b2&amp;lang.css"></script>  <script type="module" src="/entry.client.ts"></script>
  <style id="web-widget:style">web-widget{display:contents}</style>
</head>
      <body>
        <header class="_header_81u6i_14"><nav><ul><li><a href="/">Home</a></li><li><a href="/about">About</a></li></ul></nav></header><main class="_main_81u6i_31"><h1>Welcome to Web Widget.</h1><p>Try to update this message in the ./routes/index@route.tsx file.</p><h2>React</h2><!--$--><web-widget import="/routes/(components)/Counter@widget.tsx" name="ReactCounter" data="{&quot;name&quot;:&quot;React Counter&quot;,&quot;start&quot;:0}" loading="lazy" rendertarget="light" recovering=""><div class="_counter_7ir4i_1"><button>−</button><span class="_count_7ir4i_1">0</span><button>+</button></div></web-widget><!--/$--><h2>Vue</h2><!--$--><web-widget import="/routes/(components)/Counter@widget.vue?as=jsx" name="VueCounter" data="{&quot;name&quot;:&quot;Vue3 Counter&quot;,&quot;start&quot;:0}" loading="lazy" rendertarget="light" recovering=""><div class="counter" data-v-a67ee2b2><button data-v-a67ee2b2>−</button><span class="count" data-v-a67ee2b2>0</span><button data-v-a67ee2b2>+</button></div></web-widget><!--/$--></main><footer class="_footer_81u6i_38"><p>This is a footer</p></footer>
        <web-widget-inspector id="web-widget:inspector" dir="#TEST_WORKSPACE_DIR#/examples/basic" keys="[&quot;Shift&quot;]">          <script type="module" src="/@fs#TEST_WORKSPACE_DIR#/packages/web-widget/dist/inspector.js"></script>
</web-widget-inspector>
      </body>
    </html>"
`;

exports[`Should match snapshot Request "/" should match snapshot: /@headers 1`] = `
Headers {
  Symbol(headers list): HeadersList {
    "cookies": null,
    Symbol(headers map): Map {
      "access-control-allow-origin" => {
        "name": "access-control-allow-origin",
        "value": "*",
      },
      "connection" => {
        "name": "connection",
        "value": "keep-alive",
      },
      "content-type" => {
        "name": "content-type",
        "value": "text/html; charset=utf-8",
      },
      "keep-alive" => {
        "name": "keep-alive",
        "value": "timeout=5",
      },
      "transfer-encoding" => {
        "name": "transfer-encoding",
        "value": "chunked",
      },
    },
    Symbol(headers map sorted): null,
  },
  Symbol(guard): "response",
  Symbol(realm): {
    "settingsObject": {},
  },
}
`;

exports[`Should match snapshot Request "/" should match snapshot: /@status 1`] = `200`;

exports[`Should match snapshot Request "/" should match snapshot: /@statusText 1`] = `"OK"`;

exports[`Should match snapshot Request "/about" should match snapshot: /about@body 1`] = `
"<!doctype html>
    <html lang="en">
      <head>
        <script type="module">
import RefreshRuntime from "/@react-refresh"
RefreshRuntime.injectIntoGlobalHook(window)
window.$RefreshReg$ = () => {}
window.$RefreshSig$ = () => (type) => type
window.__vite_plugin_react_preamble_installed__ = true
</script>

        <script type="module" src="/@vite/client"></script>

        <script id="shim:es-module">
  if (!HTMLScriptElement.supports || !HTMLScriptElement.supports('importmap')) {
    self.importShim = Object.assign(
      function importShimProxy() {
        const esModuleShimUrl =
          'https://ga.jspm.io/npm:es-module-shims@1.7.3/dist/es-module-shims.js';
        const promise = new Promise((resolve, reject) => {
          document.head.appendChild(
            Object.assign(document.createElement('script'), {
              src: esModuleShimUrl,
              crossorigin: 'anonymous',
              async: true,
              onload() {
                if (!importShim.$proxy) {
                  resolve(importShim);
                } else {
                  reject(
                    new Error(
                      'No globalThis.importShim found:' + esModuleShimUrl
                    )
                  );
                }
              },
              onerror(error) {
                reject(error);
              },
            })
          );
        });
        return promise.then((importShim) => importShim(...arguments));
      },
      {
        $proxy: true,
      }
    );
  }
</script> <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1.0" /><title >About - Web Widget</title><script id="state:web-router" type="application/json">{"pathname":"/about","params":{},"body":{}}</script>
      <link rel="stylesheet" href="/routes/(components)/BaseLayout.module.css" />  <script type="module" src="/entry.client.ts"></script>
</head>
      <body>
        <header class="_header_81u6i_14"><nav><ul><li><a href="/">Home</a></li><li><a href="/about">About</a></li></ul></nav></header><main class="_main_81u6i_31"><h1>About</h1><p>The page you&#x27;re looking at is purely static HTML, with no client-side interactivity needed. Because of that, we don&#x27;t need to load any JavaScript. Try viewing the page&#x27;s source, or opening the devtools network panel and reloading.</p></main><footer class="_footer_81u6i_38"><p>This is a footer</p></footer>
      </body>
    </html>"
`;

exports[`Should match snapshot Request "/about" should match snapshot: /about@headers 1`] = `
Headers {
  Symbol(headers list): HeadersList {
    "cookies": null,
    Symbol(headers map): Map {
      "access-control-allow-origin" => {
        "name": "access-control-allow-origin",
        "value": "*",
      },
      "connection" => {
        "name": "connection",
        "value": "keep-alive",
      },
      "content-type" => {
        "name": "content-type",
        "value": "text/html; charset=utf-8",
      },
      "keep-alive" => {
        "name": "keep-alive",
        "value": "timeout=5",
      },
      "transfer-encoding" => {
        "name": "transfer-encoding",
        "value": "chunked",
      },
    },
    Symbol(headers map sorted): null,
  },
  Symbol(guard): "response",
  Symbol(realm): {
    "settingsObject": {},
  },
}
`;

exports[`Should match snapshot Request "/about" should match snapshot: /about@status 1`] = `200`;

exports[`Should match snapshot Request "/about" should match snapshot: /about@statusText 1`] = `"OK"`;
